tests/testthat/test-l_pkconc.R

```diff
@@ -0,0 +1,212 @@
+  TRT01A = c("Treatment1", "Treatment1", "Treatment2", "Treatment2"),
+  USUBJID = c("1", "2", "3", "4"),
+  AVISIT = c("Visit1", "Visit1", "Visit2", "Visit2"),
+  NFRLT = c(0, 1, 0, 1),
+  AFRLT = c(0, 1.12, 0, 1.055),
+  AVAL = c(0, 20.12, 30.55, NA),
+  AVALU = "mg/L",
+  stringsAsFactors = FALSE
+)
+
+# Define labels
+attr(adpc[["AFRLT"]], "label") <- "Actual time from first dose"
+attr(adpc[["NFRLT"]], "label") <- "Planned time from first dose"
+attr(adpc[["AVAL"]], "label") <- "Analysis value"
+attr(adpc[["PARAM"]], "label") <- "Analyte"
+attr(adpc[["TRT01A"]], "label") <- "Actual treatment"
+attr(adpc[["PCSPEC"]], "label") <- "Specimen"
+attr(adpc[["ROUTE"]], "label") <- "Administration"
+attr(adpc[["USUBJID"]], "label") <- "Unique Subject ID"
+attr(adpc[["AVISIT"]], "label") <- "Actual visit"
+
+describe("l_pkconc", {
+  it("creates listings for each unique combination of grouping variables", {
+    # For 1 case
+    listings_1l <- l_pkconc(adpc, listgroup_vars = c("PCSPEC"),
+                            grouping_vars = c("TRT01A", "USUBJID", "AVISIT"),
+                            displaying_vars = c("NFRLT", "AFRLT", "AVAL"))
+    expect_length(listings_1l, length(unique(adpc$PCSPEC)))
+    expect_named(listings_1l, expected = as.character(unique(adpc$PCSPEC)))
+
+    # For 2 case
+    listings_2l <- l_pkconc(adpc, listgroup_vars = c("PARAM", "PCSPEC", "ROUTE"),
+                            grouping_vars = c("TRT01A", "USUBJID", "AVISIT"),
+                            displaying_vars = c("NFRLT", "AFRLT", "AVAL"))
+    expect_length(listings_2l, length(unique(interaction(adpc$PARAM, adpc$PCSPEC, adpc$ROUTE))))
+    expect_named(listings_2l, expected = as.character(
+      unique(
+        interaction(adpc$PARAM, adpc$PCSPEC, adpc$ROUTE)
+      )
+    )
+    )
+
+    # For 4 case
+    listings_4l <- l_pkconc(adpc, listgroup_vars = c("PARAM", "PCSPEC", "ROUTE", "USUBJID"),
+                            grouping_vars = c("TRT01A", "USUBJID", "AVISIT"),
+                            displaying_vars = c("NFRLT", "AFRLT", "AVAL"))
+
+    # Include lengths and names expected
+    expect_length(listings_4l, length(unique(interaction(adpc$PARAM,
+                                                         adpc$PCSPEC,
+                                                         adpc$ROUTE,
+                                                         adpc$USUBJID))))
+    expect_named(listings_4l, expected = as.character(interaction(adpc$PARAM,
+                                                                  adpc$PCSPEC,
+                                                                  adpc$ROUTE,
+                                                                  adpc$USUBJID)))
+
+    # All in correct format (list) and each internal element internally inherits listing_df
+    expect_true(class(listings_1l) == "list" &&
+                  class(listings_2l) == "list" &&
+                  class(listings_4l) == "list")
```

Comment:
issue: Test each case separately. If one of those fails, you will not know which one exactly - only that the expectation as a whole failed.
suggestion: You can use `purrr::walk()` to streamline tests like this.
```suggestion
purrr::walk(list(listings_1l, listings_2l, listings_4l), \(x) expect_true(class(x) == "list"))
```

This comment applies to several places in the test file.